packages/ember-data/lib/serializers/json-serializer.js

```diff
@@ -274,11 +276,42 @@ export default Serializer.extend({
   normalizeErrors: function(typeClass, hash) {
     this.normalizeId(hash);
     this.normalizeAttributes(typeClass, hash);
-    this.normalizeRelationships(typeClass, hash);
+    this.normalizeRelationshipErrors(typeClass, hash);
     this.normalizeUsingDeclaredMapping(typeClass, hash);
   },
 
   /**
+    @method normalizeRelationshipErrors
+    @private
+  */
+  normalizeRelationshipErrors: function(typeClass, hash) {
+    if (!this.keyForRelationship) {
+      return;
+    }
+
+    typeClass.eachRelationship(function(key, relationship) {
+      var payloadKey = this.keyForRelationship(key, relationship.kind, 'deserialize');
+      if (key !== payloadKey && hash.hasOwnProperty(payloadKey)) {
+        hash[key] = hash[payloadKey];
+        delete hash[payloadKey];
+      }
+
+      if (!hash || !hash[key]) {
+        return;
+      }
+      if (relationship.kind === 'hasMany' && isArray(hash[key])) {
+        forEach.call(hash[key], function(errors) {
+          if (errors) {
+            this.normalizeErrors(relationship.type, errors);
```

Comment:
`this` refers to the serializer of the parent; if I am not mistaken `store.serializerFor(relationship.type).normalizeErrors` should be used here, so the correct serializer is used.

Also, I think we should add an assertion/warning for the case when the length of the errors array is not the same length of the array of records, which would indicate that there are possible errors missing.